hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/metadata/SparkHoodieBackedTableMetadataWriter.java

```diff
@@ -173,24 +174,30 @@ protected HoodieData<HoodieRecord> getFunctionalIndexRecords(List<Pair<String, F
     //       HUDI-6994 will address this.
     String columnToIndex = indexDefinition.getSourceFields().get(0);
     SQLContext sqlContext = sparkEngineContext.getSqlContext();
-
-    // Group FileSlices by partition
-    Map<String, List<FileSlice>> partitionToFileSlicesMap = partitionFileSlicePairs.stream()
-        .collect(Collectors.groupingBy(Pair::getKey, Collectors.mapping(Pair::getValue, Collectors.toList())));
-    HoodieFunctionalIndex<Column, Column> functionalIndex =
-        new HoodieSparkFunctionalIndex(indexDefinition.getIndexName(), indexDefinition.getIndexFunction(), indexDefinition.getSourceFields(), indexDefinition.getIndexOptions());
-    List<HoodieRecord> allRecords = new ArrayList<>();
-    for (Map.Entry<String, List<FileSlice>> entry : partitionToFileSlicesMap.entrySet()) {
-      String partition = entry.getKey();
-      List<FileSlice> fileSlices = entry.getValue();
-      List<HoodieRecord> recordsForPartition = Collections.emptyList();
-      if (indexDefinition.getIndexType().equalsIgnoreCase(PARTITION_NAME_COLUMN_STATS)) {
-        recordsForPartition = getFunctionalIndexRecordsUsingColumnStats(metaClient, readerSchema, fileSlices, partition, functionalIndex, columnToIndex, sqlContext);
-      } else if (indexDefinition.getIndexType().equalsIgnoreCase(PARTITION_NAME_BLOOM_FILTERS)) {
-        recordsForPartition = getFunctionalIndexRecordsUsingBloomFilter(metaClient, readerSchema, fileSlices, partition, functionalIndex, columnToIndex, sqlContext, metadataWriteConfig);
-      }
-      allRecords.addAll(recordsForPartition);
-    }
+    ForkJoinPool customThreadPool = new ForkJoinPool(parallelism);
+    List<HoodieRecord> allRecords = customThreadPool.submit(() ->
```

Comment:
Can't use engine context parallelism, because we need the context for dataset creation on which spark functions can be applied. So, going with usual java parallelism.